File: app/src/main/java/com/gft/testegft/MainActivity.java
package com.gft.testegft;

import android.content.Context;
import android.graphics.Rect;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import androidx.databinding.DataBindingUtil;

import com.gft.testegft.base.BaseActivity;
import com.gft.testegft.base.BaseViewModel;
import com.gft.testegft.databinding.ActivityMainBinding;

public class MainActivity extends BaseActivity {

    private ActivityMainBinding binding;

    private MainViewModel viewModel;

    @Override
    protected int layoutRes() {
        return R.layout.activity_main;
    }

    @Override
    protected BaseViewModel viewModel() {
        if(viewModel == null)
            viewModel = new MainViewModel();
        return viewModel;
    }

    @Override
    protected void attachViewModel() {
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.setLifecycleOwner(this);
        binding.setViewModel(viewModel);
    }

    @Override
    protected void observe() {
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            View v = getCurrentFocus();
            if (v == binding.textInputPassword) {
                viewModel.validatePassword();
            }
        }

        return super.dispatchTouchEvent(event);
    }
}

File: app/src/main/java/com/joaoneto/testeandroidv2/mainScreen/view/activity/MainActivity.kt
package com.joaoneto.testeandroidv2.mainScreen.view.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.appbar.AppBarLayout
import com.joaoneto.testeandroidv2.R
import com.joaoneto.testeandroidv2.databinding.ActivityMainBinding
import com.joaoneto.testeandroidv2.loginScreen.model.UserAccountModel
import com.joaoneto.testeandroidv2.loginScreen.ui.view.activity.LoginActivity
import com.joaoneto.testeandroidv2.mainScreen.model.StatementModel
import com.joaoneto.testeandroidv2.mainScreen.view.adapter.StatementsAdapter
import com.joaoneto.testeandroidv2.mainScreen.viewModel.MainViewModel
import com.joaoneto.testeandroidv2.util.system.Formatter
import com.joaoneto.testeandroidv2.util.system.SnackbarHelper
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private lateinit var viewModel: MainViewModel


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = DataBindingUtil
            .setContentView<ActivityMainBinding>(this, R.layout.activity_main)
        val userAccount = intent.extras?.getSerializable("userAccountData") as UserAccountModel
        binding.userAccount = userAccount
        binding.formatter = Formatter()

        viewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        setUpAppBar()
        setUpView()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        logout()

    }


    private fun setUpView() {
        logoutImageView.setOnClickListener {
            logout()
        }

        viewModel.getStatements().observe(this, Observer {

            if (it == null) {
                SnackbarHelper.message(
                    mainConstraint,
                    "Não foi possivel buscar as transações, tente novamente mais tarde"
                )
            } else {
                it.statementList?.let { statementList -> setUpRecyclerView(statementList) }
            }
        })

    }

    private fun setUpAppBar() {
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(false)
        toolbar.setNavigationOnClickListener {
        }
        collapsingToolbar.setCollapsedTitleTextColor(
            ContextCompat.getColor(
                this,
                R.color.colorWhite
            )
        )
        appBar.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { appBarLayout, verticalOfSet ->
            if (verticalOfSet == 0) {
                appbarConstraint.visibility = View.VISIBLE
                collapsingToolbar.title = " "
            } else {
                appbarConstraint.visibility = View.INVISIBLE
                collapsingToolbar.title = "Bank"

            }
        })
    }

    private fun setUpRecyclerView(statements: List<StatementModel>) {
        operationsRecyclerView.adapter =
            StatementsAdapter(
                statements
            )
        operationsRecyclerView.layoutManager =
            LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
    }


    private fun logout() {
        val intent = Intent(this, LoginActivity::class.java)
        startActivity(intent)
        finish()
    }


}
